rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeper.java

```diff
@@ -69,24 +114,64 @@ public BookKeeper(Configuration conf)
             log.error(String.format("Could not fetch Metadata for %s : %s", remotePath, Throwables.getStackTraceAsString(e)));
             throw new TException(e);
         }
-
         endBlock = setCorrectEndBlock(endBlock, fileLength, remotePath);
+        List<Location> blocksInfo = new ArrayList<>((int) (endBlock - startBlock));
+        int blockSize = CacheConfig.getBlockSize(conf);
 
-        List<Boolean> blocksInfo = new ArrayList<Boolean>((int) (endBlock - startBlock));
         for (long blockNum = startBlock; blockNum < endBlock; blockNum++) {
             totalRequests++;
+            long split = (blockNum * blockSize) / splitSize;
+
             if (md.isBlockCached(blockNum)) {
-                blocksInfo.add(true);
+                blocksInfo.add(Location.CACHED);
                 cachedRequests++;
             }
             else {
-                blocksInfo.add(false);
+                if (localSplits.contains(split)) {
+                    blocksInfo.add(Location.LOCAL);
+                    remoteRequests++;
+                }
+                else {
+                    blocksInfo.add(Location.NON_LOCAL);
+                }
             }
         }
 
         return blocksInfo;
     }
 
+    private void initializeClusterManager(int clusterType)
+    {
+        if (clusterManager == null || currentNodeIndex == -1) {
+            synchronized (lock) {
```

Comment:
add this `if (clusterManager == null || currentNodeIndex == -1)' again after acquiring lock because some other thread could have already initialized clusterManager